### ISSUE TITLE
[BUG] gypfile:false no more honored

### ISSUE THREAD

**Issue by adnan-kamili** [npm/cli#3341]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

We ship the compiled node addons with our npm package. To prevent npm from trying to build the package on npm install we set `"gypfile":false` in the package.json. It worked fine up to npm v6.x but doesn't work in npm v7.13.0.

If this is the new default behaviour and `"gypfile":false` is dropped altogether (like yarn doesn't honour this too), do we have any alternate way to prevent npm from building the package on npm install.

### Expected Behavior

npm install should not attempt to rebuild the package when "gypfile":false

### Environment

- OS: macOS 11.0
- Node: 16
- npm: 7.13.0

**Comment by nlf** [comment]
does your package specify an `install` (or `preinstall`) script that runs node-gyp?

looking at the code [here](https://github.com/npm/run-script/blob/master/lib/run-script-pkg.js#L35-L40) if you have an install or preinstall script already specified, we will run them. we won't run an implicit `node-gyp rebuild` if you do not have an install or preinstall script and you specify `"gypfile": false` though.

i think this is a bug either way, but i'd like to confirm my assumption so we can be sure we fix this correctly.

**Comment by adnan-kamili** [comment]
Here is the repo where the package exists:

https://github.com/cryptlex/lexactivator-js

To replicate the issue run ```npm install``` in the following sample:

https://github.com/cryptlex/lexactivator-js/tree/master/examples

It works fine in the case of npm 6.x but fails in npm 7.x.

**Comment by octopusjames** [comment]
I had the same issue and seems both 6.x and 7.x can't work. (version 6.14.11 and 7.16.0.)

Here is my package.json
```
{
    ...
    "gypfile": false,
    "scripts": {
        ...
        "preinstall": "node ./scripts/preinstall.js",  // just a placeholder to log something to console 
        ...
    },
    ...
}
```

---
OS: Windows 10 pro 20H2
Node: 14.16.0
npm:  6.14.11 and 7.16.0

---
Btw, is there any other workaround to disable node-gyp rebuild when npm install?
Many thanks!

**Comment by ewanharris** [comment]
@nlf I don't think this is coming from `@npmcli/run-script` but instead from `@npmcli/arborist`. 

I looked at my install process under the debugger and tracked it down to [here](https://github.com/npm/arborist/blob/ff35994c7a2ac9cb88e6c27f0e9a45743a611c35/lib/arborist/rebuild.js#L230), at this point the `pkg` variable is a reduced version of my dependencies package.json, and is actually missing the `scripts` stanza (and also the `gypfile` like the original issue stated) so is using the default `node-gyp rebuild` always.

I forced it to take [this path](https://github.com/npm/arborist/blob/ff35994c7a2ac9cb88e6c27f0e9a45743a611c35/lib/arborist/rebuild.js#L210-L225) and refresh the package.json data and the install worked.

**Comment by ewanharris** [comment]
I was able to workaround this for now by manually reinstating the `"hasInstallScript": true` property onto the dependency in the package-lock.json file. It appears this had been removed while installing a new dependency causing the above, [ref](https://github.com/appcelerator/titanium_mobile/pull/12873/commits/5a174ce31a93838d79562e5c3faed45235245734)

**Comment by kgryte** [comment]
We're also wanting to disable native add-on compilation by setting `gypfile: false` in `package.json` files. Unfortunately, I cannot find anything in the `npm` docs explaining the expected behavior and whether `gypfile: true|false` is honored. Would be great if the public `npm` docs could explicitly clarify what should happen (preferably in the [`package.json` docs](https://docs.npmjs.com/cli/v7/configuring-npm/package-json)).

**Comment by Planeshifter** [comment]
Following up on this, while most recent versions of `npm` on Linux do honor `"gypfile": false` in `package.json` files, we and our [users](https://github.com/stdlib-js/stdlib/issues/465) have encountered aforementioned problem on Windows 10/11 with the latest `npm` versions `8.1.x` down to `7.6.3`. It would be great to get some clarification on whether the expected behavior of the `gypfile` field is intentionally left vague or whether the reported behavior will be treated as a bug.

**Comment by rotu** [comment]
~~I think this issue should be closed as fixed by https://github.com/npm/cli/commit/281a7f39ac314bd7657ce2bcd7918b21eee99210.~~ @Planeshifter is this still an issue on Windows? I opened a new issue for the missing `package.json` docs #6898.
